calyx-py/test/correctness/queues/binheap/binheap.py

```diff
@@ -203,56 +204,90 @@ def extract_snd(name, idx, out):
         le_1.left = size.out
         le_1.right = child_r_idx.out
 
-        le_2.left = child_l_rank.out
-        le_2.right = child_r_rank.out
+        le_2.left = child_r_idx.out
+        le_2.right = current_idx.out
+
+        le_3.left = child_l_rank.out
+        le_3.right = child_r_rank.out
 
-        if_or.left = le_1.out
-        if_or.right = le_2.out
+        inner_or.left = le_1.out
+        inner_or.right = le_2.out
 
-    lt_1 = comp.lt(queue_len_factor)
-    lt_2 = comp.lt(32)
-    lt_3 = comp.lt(queue_len_factor)
-    lt_4 = comp.lt(32)
-    and_1 = comp.and_(1)
-    and_2 = comp.and_(1)
+        if_or.left = inner_or.out
+        if_or.right = le_3.out
+
+    lt_l_1 = comp.lt(addr_size)
+    lt_l_2 = comp.lt(addr_size)
+    lt_l_3 = comp.lt(64)
+    lt_r_1 = comp.lt(addr_size)
+    lt_r_2 = comp.lt(addr_size)
+    lt_r_3 = comp.lt(64)
+    and_l_1 = comp.and_(1)
+    and_l_2 = comp.and_(1)
+    and_r_1 = comp.and_(1)
+    and_r_2 = comp.and_(1)
     while_or = comp.or_(1)
     with comp.comb_group("current_gt_children") as current_gt_children:
-        # Check if the `current_idx`th element should be swapped with its left OR right.
-        # child_l_idx < size AND child_l_rank < current_rank
+        # Check if the `current_idx`th element should be swapped with its left OR right child.
+        # child_l_idx < size AND current_idx < child_l_idx AND child_l_rank < current_rank
         # OR
-        # child_r_idx < size AND child_r_rank < current_rank
-        lt_1.left = child_l_idx.out
-        lt_1.right = size.out
+        # child_r_idx < size AND current_idx < child_r_idx AND child_r_rank < current_rank
+        lt_l_1.left = child_l_idx.out
+        lt_l_1.right = size.out
+
+        lt_l_2.left = current_idx.out
+        lt_l_2.right = child_l_idx.out
 
-        lt_2.left = child_l_rank.out
-        lt_2.right = current_rank.out
+        lt_l_3.left = child_l_rank.out
+        lt_l_3.right = current_rank.out
 
-        lt_3.left = child_r_idx.out
-        lt_3.right = size.out
+        lt_r_1.left = child_r_idx.out
+        lt_r_1.right = size.out
 
-        lt_4.left = child_r_rank.out
-        lt_4.right = current_rank.out
+        lt_r_2.left = current_idx.out
+        lt_r_2.right = child_r_idx.out
 
-        and_1.left = lt_1.out
-        and_1.right = lt_2.out
+        lt_r_3.left = child_r_rank.out
+        lt_r_3.right = current_rank.out
 
-        and_2.left = lt_3.out
-        and_2.right = lt_4.out
+        and_l_1.left = lt_l_1.out
+        and_l_1.right = lt_l_2.out
+        and_l_2.left = and_l_1.out
+        and_l_2.right = lt_l_3.out
 
-        while_or.left = and_1.out
-        while_or.right = and_2.out
+        and_r_1.left = lt_r_1.out
+        and_r_1.right = lt_r_2.out
+        and_r_2.left = and_r_1.out
+        and_r_2.right = lt_r_3.out
+
+        while_or.left = and_l_2.out
+        while_or.right = and_r_2.out
```

Comment:
Wild! I'm guessing the bug was somewhere in here?

Reply:
There were multiple bugs, but I think you're right that one of them lived here: the issue where my PIFO and FIFO tests didn't terminate.

Before, to verify if a left or right child existed, I'd check if `child_idx < size`. This generally works fine when the heap is large enough. Once we start putting values in the top half of the array, an additional check to ensure `child_idx < current_idx` is important, I think (since `2 * current_idx` can now loop back around). These changes introduced that check.